src/lib/car-index.js

```diff
@@ -147,3 +151,71 @@ export class StreamingCarIndex {
 }
 
 const mhToKey = (/** @type {Uint8Array} */ mh) => base58btc.encode(mh)
+
+export class BlocklyIndex {
+  /** R2 bucket where indexes live. */
+  #bucket
+  /** Cached index entries. */
+  #cache
+  /** Indexes that have been read. */
+  #indexes
+
+  /**
+   * @param {import('../bindings').R2Bucket} indexBucket
+   */
+  constructor (indexBucket) {
+    this.#bucket = indexBucket
+    /** @type {Map<RawCIDString, IndexEntry>} */
+    this.#cache = new Map()
+    this.#indexes = new Set()
+  }
+
+  /** @param {UnknownLink} cid */
+  async get (cid) {
+    // console.log(`index get: ${cid}`)
+    const key = cidToKey(cid)
+    let indexItem = this.#cache.get(key)
+    if (indexItem) {
+      // console.log(`index cache HIT: ${indexItem.origin}: ${cid} @ ${indexItem.offset}`)
+      if (cid.code !== raw.code) {
+        await this.#readIndex(cid)
+      }
+    } else {
+      await this.#readIndex(cid)
+      indexItem = this.#cache.get(key)
+      if (!indexItem) return // weird huh!?
+    }
+    return { cid, ...indexItem }
+  }
+
+  /**
+   * @param {import('multiformats').UnknownLink} cid
+   */
+  async #readIndex (cid) {
+    const key = cidToKey(cid)
```

Comment:
nit: could be nice to name this key as `indexesKey`. I initially thought it was `bucketKey`. But then saw it was not.